base/system/services/database.c

```diff
@@ -852,6 +852,105 @@ ScmDeleteServiceRecord(PSERVICE lpService)
     DPRINT("Done\n");
 }
 
+DWORD
+Int_EnumDependentServicesW(HKEY hServicesKey,
+                           PSERVICE lpService,
+                           DWORD dwServiceState,
+                           PSERVICE *lpServices,
+                           LPDWORD pcbBytesNeeded,
+                           LPDWORD lpServicesReturned);
+
+DWORD ScmDeleteService(PSERVICE lpService)
+{
+    DWORD dwError;
+    DWORD pcbBytesNeeded = 0;
+    DWORD dwServicesReturned = 0;
+    HKEY hServicesKey;
+
+    ASSERT(lpService->RefCount == 0);
+
+    /* Open the Services Reg key */
+    dwError = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                            L"System\\CurrentControlSet\\Services",
+                            0,
+                            KEY_SET_VALUE | KEY_READ,
+                            &hServicesKey);
+    if (dwError != ERROR_SUCCESS)
+    {
+        DPRINT("Failed to open services key\n");
+        return dwError;
+    }
+
+    /* Call the internal function with NULL, just to get bytes we need */
+    Int_EnumDependentServicesW(hServicesKey,
+                               lpService,
+                               SERVICE_ACTIVE,
+                               NULL,
+                               &pcbBytesNeeded,
+                               &dwServicesReturned);
+
+    /* If pcbBytesNeeded returned a value then there are services running that are dependent on this service */
+    if (pcbBytesNeeded)
+    {
+        DPRINT("Deletion failed due to running dependencies.\n");
```

Comment:
See comment https://github.com/reactos/reactos/pull/1225/files#r280022759

Reply:
In the current state of ReactOS, it does very minimal handling of the services dependencies at all. It does somewhat show them in the Properties | Dependencies tab, but that is about all.

![Services_Dependencies-01](https://github.com/user-attachments/assets/b65d9cb6-0130-40fc-b34d-812391000f87)

You will notice that the description for the lower treeview window is completely missing its description.
This should be 'The following components depend on this service:' or similar and I intend to make a PR soon.
At this time there is no dependency tracking anywhere else in the code, so this deserves quite a bit more attention before the adding code as suggested in the comment would provide any benefit. Thanks.

Reply:
Yes, this is indeed a can of worms on its own. For the record: My testing on WinXPSP3 with XPSP3's own services.exe and XPSP3s own services.msc and XPSP3s own set of MS-stock-services, I do also know a few cases where services are having cross-dependencies, and still XPSP3 does not know how to **start** them in the right order, even when the services are being properly marked as having a dependency-chain. What always seemed to work though for my on XPSP3 is when the user manually stopped the services from the top-to-down in the correct order. It most of the times also seems to work (most of the time) to stop a service, and at the same time stopping the services that rely on the other one as well. XPSP3 aks then something like "Stopping Service x will stop Service y and z as well. Do you want to continue?".

But contrary to that, there are some services that are impossible to be restarted again, even when the user tries to manually start them in the correct order from the bottom to the top of the dependency graph.

Long story short: The task being discussed here seems to be so complicated in some cases, that even MS didn't manage to get it 100% correct.

For now in ReactOS some services, no matter whether they do have any dependencies on top of them or to the bottom, can simply not be stopped properly. And if Dougs PR would be able to fix that, this would be a huge gain already. Let's leave the complicated things for a future PR.

**But: Let's turn that DPRINT into a DPRINT1 at least!**

Reply:
Changed DPRINT to DPRINT1 in database.c Line 895 as requested. Thanks.

Reply:
It may be interesting to note that currently ReactOS does not even offer the "Stop" option on the Remote Procedure Call service when right-clicked, but in previous revisions this was still possible. When attempted, there was a popup warning displayed as shown here.

![Stop_Other_Services_Warning](https://github.com/user-attachments/assets/9357cdbd-a0a2-4dd0-8467-a802cba4f00e)

So that code is most likely still present to provide this. Thanks.

Reply:
Thx for mentioning that detail. I wasn't even aware of that regression yet. It deserves an individual JIRA-ticket and should be bisected as well. I am not sure whether the guilty rev for that will turn out to be 0.4.9-dev-60-g1dfbed9 https://github.com/reactos/reactos/commit/1dfbed9c3dbd135c7bf51090b9cadde6d2642340 once more (same as it was for https://jira.reactos.org/browse/CORE-16949 or whether that question-box broke by any other revision instead. I will file a JIRA-ticket so it won't be forgotten!

Reply:
> It may be interesting to note that currently ReactOS does not even offer the "Stop" option on the Remote Procedure Call service when right-clicked, but in previous revisions this was still possible. When attempted, there was a popup warning displayed as shown here.

The RPC service is marked as non stoppable, so why would you want the stop button to be enabled?

Reply:
I wrote that already. **This discussion is resolved already.** But I do think MS was overcautious here. I do know that the unstoppable nonsense can be commented out in the registry, or be circumvented by using the taskmgr to kill it. Or by using an additional reboot after having manipluated that services startup-type.

Reply:
Note that you could also test with command line tool `sc`